### ISSUE TITLE
Implement neutral exit status on bin/filter

### ISSUE THREAD

**Issue by swinton** [actions/bin#13]
[GitHub Actions can now have neutral status](https://developer.github.com/actions/changes/2018-10-24-github-action-exit-codes/), to gracefully exit a workflow early.

We should update [the `bin/filter` Action](https://github.com/actions/bin/tree/master/filter) to take advantage of [this functionality](https://developer.github.com/actions/creating-github-actions/accessing-the-runtime-environment/#exit-codes-and-statuses), triggering a neutral state by exiting with status code `78`.

**Comment by NiklasMerz** [comment]
Looks like 0ac6d4416cc1ef450ef4cf5b992ae6e085ba3024 already fixed this. But I am trying to find out why my actions after the filter get the status cancelled. Is this by design?

**Comment by JoshuaToenyes** [comment]
@NiklasMerz I'm seeing the same thing... 

[From the docs](https://developer.github.com/actions/creating-github-actions/accessing-the-runtime-environment/#exit-codes-and-statuses):

> The associated check run shows a neutral status, and the overall check suite will have a status of success as long as there were no failed or **cancelled** actions.

But I'm seeing that follow-on actions are **cancelled**, which would make the suite fail... which doesn't make sense to me. Seems that follow on actions should also have a neutral status, since they weren't ever started. Which leads me to ask the same question, by design? Anyone from GitHub have insight on this?

**Comment by JoshuaToenyes** [comment]
Until we have a long-term resolution on this from GitHub, I've simply pulled-in some of GitHub's filter action login into my critical action.

```sh
pattern="refs/heads/${BRANCH_FILTER:-*}"
filter "$pattern"
ret=$?

if [ $ret -ne 0 ]; then
  exit $ret
fi
```

[Relevant code in my `entrypoint.sh`](https://github.com/RobotSnacks/actions/blob/0d1f38733d09feb25f1bd732d29f6902169d12d1/lerna/entrypoint.sh#L20-L26)

The above is simply using a [copy of GitHub's `ref` filter shell script](https://github.com/RobotSnacks/actions/blob/0d1f38733d09feb25f1bd732d29f6902169d12d1/lerna/filter.sh), which I've renamed `filter`, then I'm using an environment variable `$BRANCH_FILTER` to filter on the branch.

Ultimately, I'm not sure if a "canceled" action should cause the whole suite to be marked with a red "X" in GitHub's UI... it's not very helpful in my use case. I'd rather see them marked as neutral.

In my case, I'm simply filtering on the branch to prevent publishing artifacts on non-master branches in a single action.

**Comment by johnnyreilly** [comment]
I'm experiencing this problem too:

It looks like filtering makes subsequent actions in GitHub show as cancelled because the filter item has failed. That seems... Not quite right? Eg:

"Action publish' cancelled because action check for new tag' failed"

https://github.com/TypeStrong/ts-loader/pull/893

**Comment by NiklasMerz** [comment]
I contacted the github support about that and they forwared the request to the engineering team.

[This](https://github.community/t5/GitHub-API-Development-and/Cancelled-Github-actions-fails-PR-checks/m-p/15998#M440) forum thread deals about that, too.

**Comment by johnnyreilly** [comment]
Would anyone be able to comment on if / when this is likely to be looked at please?  For now I've commented out the part of my workflow that depends upon filtering, hoping to introduce it when this issue is remedied.  

I've been tempted to give @JoshuaToenyes workaround a shot but ultimately I'd rather use the official workflow.  

There's no particular urgency; obviously GitHub Actions are in beta. But it would be cool to know if there's a desire to look at this at some point.

**Comment by NiklasMerz** [comment]
You could check out my "workaround"? When I build my workflow like that the status gets set to neutral.
Just use filters in resolving actions?

https://github.community/t5/GitHub-API-Development-and/Cancelled-Github-actions-fails-PR-checks/m-p/17464/highlight/true#M639

**Comment by johnnyreilly** [comment]
That's interesting @NiklasMerz - thanks for sharing!  DId you hear back from @mcolyer as to whether this is how it's supposed to be used?

**Comment by mcolyer** [comment]
@johnnyreilly thanks for the ping. I'm trying to reproduce what others are alluding to here. Given this workflow (and a push in a branch that isn't labeled with `test`):

```hcl
workflow "Verify ps" {
  resolves = ["ps1"]
  on = "pull_request"
}

action "filter" {
  uses = "actions/bin/filter@master"
  runs = "filter label test"
}

action "ps1" {
  needs = "filter"
  uses = "mcolyer-test/actions-test@master"
}
```

I see this:

![image](https://user-images.githubusercontent.com/1424/51280287-23eb7b00-1994-11e9-8122-84047bc84d03.png)

Given the above, it looks like it's working as expected - `ps1` is marked as `neutral` and not cancelled. @NiklasMerz, @JoshuaToenyes or @johnnyreilly what am I missing?

**Comment by johnnyreilly** [comment]
It doesn't look like what I and others have experienced.

You can take a look at my own experience here: https://github.com/TypeStrong/ts-loader/pull/893

Hopefully you can see from the last 2 commits and runs in azure pipelines what happened when I tried.

If it helps I'd be happy to try and reproduce it again on a fresh pr to make your life easier?

**Comment by balassit** [comment]
Similar issue for me. I want to skip the docker login and deploy when the branch is not master, but all of my actions are marked as neutral. The `Build` action should be completed because it not dependent of the filter
```
workflow "Docker Deploy" {
  resolves = [
    "Deploy to Docker"
  ]
  on = "push"
}

action "filter-to-branch-master" {
  uses = "actions/bin/filter@master"
  args = "branch master"
}

action "Build" {
  uses = "actions/docker/cli@76ff57a6c3d817840574a98950b0c7bc4e8a13a8"
  runs = "docker build -t balassit/my-angular-project:prod ."
}

action "Docker Login" {
  uses = "actions/docker/login@76ff57a6c3d817840574a98950b0c7bc4e8a13a8"
  needs = ["filter-to-branch-master"]
  secrets = ["DOCKER_USERNAME", "DOCKER_PASSWORD"]
}

action "Deploy to Docker" {
  uses = "actions/docker/cli@76ff57a6c3d817840574a98950b0c7bc4e8a13a8"
  needs = ["Docker Login", "Build"]
  runs = "docker push balassit/my-angular-project:prod"
}
```
<img width="718" alt="Github action" src="https://user-images.githubusercontent.com/10055646/51301295-8d46aa80-19e3-11e9-8bde-af8124ac5631.png">

**Comment by nicolo-ribaudo** [comment]
@balassit It doesn't seem totally unreasonable: since the `Deploy to Docker` action is stopped, all it's children can stopped. Does it work if you also add `Build` in `Docker Deploy`'s `resolves`?

**Comment by balassit** [comment]
@nicolo-ribaudo assuming you meant adding `Build` and `Docker Login` like this:
```
workflow "Docker Deploy" {
  resolves = [
    "Deploy to Docker",
    "Build",
    "Docker Login"
  ]
  on = "push"
}
```
I get the same result. Looking at the `Build` log, it says
> ### SKIPPED Build 07:45:46Z
> Action 'Build' skipped because action `filter-to-branch-master' failed

**Comment by mcolyer** [comment]
> Similar issue for me. I want to skip the docker login and deploy when the branch is not master, but all of my actions are marked as neutral. The `Build` action should be completed because it not dependent of the filter

Ahh so I think I know what's going on here now. So this is a known issue that we're working to correct. If you have multiple branches in your workflow, currently as soon as one of has an action that stops all of the branches stop. In the example above, we want the `build` action to execute regardless of the result of the `filter-to-branch-master` action and are working on a fix.

**Comment by johnnyreilly** [comment]
@mcolyer it sounds like there's 2 issues being discussed in this thread as the ts-loader case has only a single branch. Did you get what you needed from the link: 
https://github.com/TypeStrong/ts-loader/pull/893

Happy to raise a new pr to illustrate the issue if that's not clear enough.

**Comment by mcolyer** [comment]
> I'm happy to raise a new PR to illustrate the filter issue around subsequent steps after a neutral being marked as failed if that's not clear enough.

That would be helpful. When I looked at it yesterday I wasn't sure if your issue was separate or not - looks like I missed it. Having a new PR would help me see exactly what you mean.

**Comment by johnnyreilly** [comment]
No worries - I'll put something together this weekend and report back.

**Comment by johnnyreilly** [comment]
The weirdest thing - it's now behaving as desired: https://github.com/TypeStrong/ts-loader/pull/898

I think we're all good!

**Comment by jlelse** [comment]
I can also confirm, since today everything seems to work as expected.

**Comment by mcolyer** [comment]
@johnnyreilly and @jlelse given that, would you be okay if I closed this issue out as I think the issue this was opened around is fixed and we're already tracking the other one internally and will communicate more broadly when it's fixed?

**Comment by Babsvik** [comment]
> Ahh so I think I know what's going on here now. So this is a known issue that we're working to correct. If you have multiple branches in your workflow, currently as soon as one of has an action that stops all of the branches stop. In the example above, we want the `build` action to execute regardless of the result of the `filter-to-branch-master` action and are working on a fix.

@mcolyer do you have an URL to the issue tracking this? Don't want to make duplicate issues, but couldn't find any issue tracking that. This is exactly the problem I'm currently experiencing 👍

**Comment by jlelse** [comment]
It seemed to be fixed but now happening again for a few days.

**Comment by Babsvik** [comment]
I wanted to find the original issue for this, because I feel this issue is targeting something else.

But I couldn't find the issues @mcolyer was refering to handling this, that is why I was asking.

To be clear my problem is:
it's stopping, when an "unrelated" check fails.

![screenshot 2019-03-02 at 03 30 31](https://user-images.githubusercontent.com/1538003/53682982-d769be00-3cfb-11e9-8826-8f947546b00c.png)

The one that is canceled on the right, is canceled because:
![screenshot 2019-03-02 at 03 30 42](https://user-images.githubusercontent.com/1538003/53683004-02eca880-3cfc-11e9-976c-7e5995b32113.png)

`"Action 'deploy-master' cancelled because action 'branch-beta-filter' failed"`

Seing as it's "not related" it does not make sense to me that it fails because of that. I could put it in another workflow, but seeing as I'm only going to deploy to one target at any time. It makes more sense to have it like this, so it does not run both in parallel.

Being pointed in the right direction would be much appreciated 👍 

Workflow code, below:

```
workflow "Deploy on Zeit" {
  on = "push"
  resolves = [
    "alias-master",
    "alias-beta",
  ]
}

# Master Pipeline

action "branch-master-filter" {
  uses = "actions/bin/filter@master"
  args = "branch master"
}

action "deploy-master" {
  needs = ["branch-master-filter"]
  uses = "actions/zeit-now@master"
  args = "deploy --local-config=./now-prod.json"
  secrets = [
    "ZEIT_TOKEN",
  ]
}

action "alias-master" {
  needs = ["deploy-master"]
  uses = "actions/zeit-now@master"
  args = "alias --local-config=./now-prod.json"
  secrets = [
    "ZEIT_TOKEN",
  ]
}

# Beta Pipeline

action "branch-beta-filter" {
  uses = "actions/bin/filter@master"
  args = "branch beta"
}

action "deploy-beta" {
  needs = ["branch-beta-filter"]
  uses = "actions/zeit-now@master"
  args = "deploy --local-config=./now-beta.json"
  secrets = [
    "ZEIT_TOKEN",
  ]
}

action "alias-beta" {
  needs = ["deploy-beta"]
  uses = "actions/zeit-now@master"
  args = "alias --local-config=./now-beta.json"
  secrets = [
    "ZEIT_TOKEN",
  ]
}
```

**Comment by jlelse** [comment]
@Babsvik You need to create to workflows for that, one for master, one for beta. Had the problem too.

**Comment by NiklasMerz** [comment]
I created to workflows, too and reused some actions in both workflows.

**Comment by Babsvik** [comment]
Thanks @jlelse and @NiklasMerz ! I got that part and I understand that, I also did create two workflows for this to work for now.

My point was that @mcolyer said they are working on a fix, if I knew where I could "follow" that issue work on fixing this. I could see when it's fixed and rewrite my workflow after it's fixed :) So I'm only trying to find the issue tracking that :)
